### ISSUE TITLE
Fold improvements discussion

### ISSUE THREAD

**Issue by teto** [neovim/neovim#5931]
I've been working to improve the terminal folding display and some things I think are LGTM:
- [x] multibyte display (so one can use '│' to replace the pipe,'▸' instead of + ▾ etc: see https://transfer.sh/s663v/2017-01-11-130542-956x613-scrot.png for an example) 
- [ ] don't repeat initial folding character on wrapped lines
- [ ] Add tests, functional/legacy/045_folding_spec.lua does not test screen output much
- [ ] with the current code it should be easy to adjust dynamically the foldcolumn width depending on the nesting fold depth

But some I am less confident with and would like some input:
- [ ] multicell display ? I wonder if some multicell character can improve the foldcolumn look enough to justify the addition. Which character would you like (I could only find that one '＋') ?
- [ ] character customization ? it's easy to change foldcolumn characters in the source but that might be a nice setting to have ? In fillchars ?
- [ ] add nvim_create_fold() to API ?
- [ ] help with folds UI externalization ? GUIs could draw better folds for sure

Nb: as a reminder for myself: lines displayed on screen are generated by win_line and fold_line. Both functions write their result in `current_ScreenLine` which is copied later at the right offset.
fold_line fills `current_ScreenLine` via `copy_text_attr` calls but this function doesn't support multibytes characters. `screen_puts_len` is a powered up version of `copy_text_attr` that support multibytes characters BUT it can't write to 'current_ScreenLine' because it can't compute its offset.

**Comment by justinmk** [comment]
> with the current code it should be easy to adjust dynamically the foldcolumn width depending on the nesting fold depth

Won't that make the screen "bounce around" when opening/closing a fold? Allowing the foldcolumn width to be configurable should be enough. Any deeper folds should just show the fold symbol in the rightmost column of the foldcolumn (leftmost when `'rtl'` is set).

> multicell display ? I wonder if some multicell character can improve the foldcolumn look enough to justify the addition. Which character would you like (I could only find that one '＋') ?

I wouldn't bother with that. If the fold display is "externalized" later, even a TUI could show a custom layout.

> character customization ? it's easy to change foldcolumn characters in the source but that might be a nice setting to have ? In fillchars ?

Given that "stl:" is part of fillchars, adding "foldcolumn:" or something like that, makes sense.

> add nvim_create_fold() to API ?

👍  Definitely in favor.

>  help with folds UI externalization ? GUIs could draw better folds for sure

I suggest waiting until #5686 is finished.

**Comment by teto** [comment]
> Won't that make the screen "bounce around" when opening/closing a fold? Allowing the foldcolumn width to be configurable should be enough. Any deeper folds should just show the fold symbol in the rightmost column of the foldcolumn (leftmost when 'rtl' is set).
What I was thinking of is:
-Opening/closing a fold doesn't change the fdc width
-Creation/fold can change fdc width and "bounce". If neovim looks at the buffer-wide maximum nesting depth, the bouncing should be relatively infrequent
For instance, in my config I have `:set fdc=3` and when my buffer has no folds, nvim loses 3 columns for nothing.

>I wouldn't bother with that. If the fold display is "externalized" later, even a TUI could show a custom layout.
guess you are right, it adds too much complexity and limits the number of folds that can be displayed.

**Comment by teto** [comment]
I got fillchars working. 
My problem with adaptive fdc is that is changes the meaning of the variable, i.e. it becomes the maximum width. As it might be controversial I will focus on improving tests first.

Btw, if anyone knows how to know if the current screen line is a wrapped line or not, it could spare me some analysis (seems that `LineWraps` is the key but does it accepted a "screenline" or "file line" ?).

**Comment by teto** [comment]
I am very tempted to change the current behavior when `current line foldlevel > fdc`, i.e., vim prints somehow the fold levels. It does not make much sense to me but I am curious if others feel different, if they like it and how they use it:
``` 
   aa
-  {
|- {
2- {
23 bb
23 }
|| cc
|| }
|  dd
|  ee
|+ +---  3 lignes : --------------------------------------------------------------------------------------------------------------------
|  }
   gg
``` 

I would like to convey the information there are hidden folds in another way.Highlights are too distracting.
I thought of some alternatives:
- display the maximum depth just on the first line of the screen or the last line.
- display the number of missing columns only to the right handmost column.  The fdc would look like `||3`. Visually I would opt for `||>`but it conveys less information and might bother some.
Compared to the previous example; somethink like
```
|- {
|- {
|3 bb
|3 }
|| cc
|| }
|  dd
```

**Comment by justinmk** [comment]
Neat idea, but might be overwrought. Users should already know what to expect based on their 'foldnestmax', for example.

**Comment by teto** [comment]
> Users should already know what to expect based on their 'foldnestmax

Agreed that's why I find the vim numbers more confusing than anything, makes current folds hard to read. Am I right in understanding you would be ok with another behavior ? In fact I finally found a behavior I am comfortable with, ie., on the last column, just do as in all other editors  and put fold open/closed on the same line. Here is a preview with a max nesting of 3.

With enough fdc, there is no problem:
![2017-01-27-214323_956x487_scrot](https://cloud.githubusercontent.com/assets/886074/22387043/1d580326-e4da-11e6-9f10-4c964b6b107c.png)

But when the number of columns is not enough to display each fold on their column, you merge all the folds with a level >= fdc on the same column; example with fdc=2:
![2017-01-27-214315_956x487_scrot](https://cloud.githubusercontent.com/assets/886074/22387041/1953a24e-e4da-11e6-90fa-b690389ee7ba.png)

...and fdc=1:
![2017-01-27-214307_956x487_scrot](https://cloud.githubusercontent.com/assets/886074/22387031/1323b86e-e4da-11e6-896b-f37cef726db2.png)

If you are ok with it I can proceed with cleaning the code and writing tests.

**Comment by justinmk** [comment]
Can you annotate the pictures? I'm not seeing how that's different from the OP or existing behavior.
